[BUG] asyncdns ForwardResolver crashes when DNS error and empty response arrive in wrong order

### Describe the bug

`Resolver._save_result` in `nav/asyncdns.py` crashes with `AttributeError: 'DNSNameError' object has no attribute 'extend'` when a forward DNS lookup produces a DNS error for one record type (e.g. AAAA) and an empty success response for the other (e.g. A), and the error callback is processed first.

`ForwardResolver.lookup()` issues two parallel lookups per name (A and AAAA). Both results are passed to `_save_result`, which assumes `self.results[name]` is always a list when the response is not an exception. If the error callback runs first, it replaces the default list with an exception object, and the subsequent successful callback tries to call `.extend()` on that exception.

There is also a silent bug in the opposite ordering: if the success callback runs first, the error callback unconditionally overwrites the valid address list with the exception.

### To Reproduce

This manifests as a flaky failure in the `ipdevinfo` host info view whenever the DNS lookup for a device's sysname triggers a `DNSNameError` on one record type before the other record type returns. Observed in CI as an intermittent failure of `tests.integration.web.ipdevinfo_test.TestHostInfoModal.test_should_render_modal` on Python 3.11.

### Expected behavior

A DNS error on one record type (e.g. AAAA) should not crash the resolver or discard valid results from the other record type (e.g. A).

### Tracebacks

```pytb
self = <tests.integration.web.ipdevinfo_test.TestHostInfoModal object at 0x7f7e03cec290>
client = <django.test.client.Client object at 0x7f7de7c08910>
netbox = <Netbox: example-sw.example.org>

    def test_should_render_modal(self, client, netbox):
        url = reverse(
            "ipdevinfo-hostinfo",
            args=[netbox.sysname],
        )
>       response = client.get(url)

tests/integration/web/ipdevinfo_test.py:246:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
python/nav/web/ipdevinfo/views.py:856: in render_host_info
    {'host_info': get_host_info(identifier)},
python/nav/web/ipdevinfo/host_information.py:56: in get_host_info
    addresses = forward_lookup(host) or []
python/nav/web/ipdevinfo/host_information.py:33: in forward_lookup
    addrinfo = asyncdns.forward_lookup([host])
python/nav/asyncdns.py:82: in forward_lookup
    return resolver.resolve(names)
python/nav/asyncdns.py:129: in resolve
    raise error
python/nav/asyncdns.py:146: in _save_result
    self.results[name].extend(response)
E   AttributeError: 'DNSNameError' object has no attribute 'extend'
```

### Environment

Observed in CI (GitHub Actions), Python 3.11, integration test suite.

### Additional context

The root cause is in `Resolver._save_result`: it doesn't account for the fact that multiple callbacks write to `self.results[name]`, and an earlier error callback can replace the default `list` with an exception object.

--- a/changelog.d/3902.fixed.md
+++ b/changelog.d/3902.fixed.md
@@ -1,0 +1,1 @@
+Fixed a race condition crash in asyncdns `ForwardResolver` where a DNS error on one record type (A/AAAA) could cause an `AttributeError` when the other record type resolved successfully.

--- a/python/nav/asyncdns.py
+++ b/python/nav/asyncdns.py
@@ -141,8 +141,9 @@ class Resolver(object):
     def _save_result(self, result):
         name, response = result
         if isinstance(response, Exception):
-            self.results[name] = response
-        else:
+            if not self.results[name]:
+                self.results[name] = response
+        elif isinstance(self.results[name], list):
             self.results[name].extend(response)
 
     @staticmethod

--- a/tests/unittests/asyncdns_test.py
+++ b/tests/unittests/asyncdns_test.py
@@ -1,7 +1,10 @@
+from collections import defaultdict
+
 from mock import Mock
 from twisted.names import dns
+from twisted.names.error import DNSNameError
 
-from nav.asyncdns import ForwardResolver
+from nav.asyncdns import ForwardResolver, Resolver
 
 BUICK_NAME = "buick.lab.uninett.no"
 
@@ -40,3 +43,38 @@ def test_forward_lookup_should_work_with_different_case_name_results():
 
     assert result_name == BUICK_NAME
     assert address_list == ["158.38.152.162"]
+
+
+class TestSaveResult:
+    """Tests for Resolver._save_result"""
+
+    def _make_resolver(self):
+        """Returns a Resolver instance without triggering __init__'s side effects"""
+        resolver = object.__new__(Resolver)
+        resolver.results = defaultdict(list)
+        return resolver
+
+    def test_when_error_arrives_after_success_it_should_keep_addresses(self):
+        resolver = self._make_resolver()
+        resolver._save_result(("host", ["10.0.0.1"]))
+        resolver._save_result(("host", DNSNameError()))
+        assert resolver.results["host"] == ["10.0.0.1"]
+
+    def test_when_success_arrives_after_error_it_should_not_raise(self):
+        resolver = self._make_resolver()
+        resolver._save_result(("host", DNSNameError()))
+        resolver._save_result(("host", []))
+        assert isinstance(resolver.results["host"], DNSNameError)
+
+    def test_when_both_succeed_it_should_combine_results(self):
+        resolver = self._make_resolver()
+        resolver._save_result(("host", ["10.0.0.1"]))
+        resolver._save_result(("host", ["::1"]))
+        assert resolver.results["host"] == ["10.0.0.1", "::1"]
+
+    def test_when_both_fail_it_should_store_first_error(self):
+        resolver = self._make_resolver()
+        first_error = DNSNameError()
+        resolver._save_result(("host", first_error))
+        resolver._save_result(("host", DNSNameError()))
+        assert resolver.results["host"] is first_error

